src/systemctl/systemctl-list-units.c

```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
         return machine ? strjoin(machine, ":", unit) : strdup(unit);
 }
 
-static int output_units_list(const UnitInfo *unit_infos, size_t c) {
+typedef struct {
+        char ** units;
+        size_t num_units;
+}SliceUnits;
+
+/* Function to insert or update units under slice unit in the hash map */
+static int insert_units_under_slice(const char *slice_name, char **units, size_t num_units) {
+        SliceUnits *slice_units = NULL;
+        int r;
+
+        /* Initialize the hashmap */
+        if (!slice_unit_map) {
+                slice_unit_map = hashmap_new(&string_hash_ops);
+                if (!slice_unit_map)
+                        return log_oom();
+        }
```

Comment:
please don't allocate ahead of time, but do this on first insertion. i.e. hashmap_ensure_put() makes that easy.